---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE, message=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
library(hlatools)
```

# hlatools

## Overview 

`hlatools` provides a collection of tools to work with [IPD-IMGT/HLA](https://www.ebi.ac.uk/ipd/imgt/hla/) data

## Installation

Install from GitHub using `remotes`:
```{r, eval=FALSE}
remotes::install_github("gschofl/hlatools")
```

## Usage

### Get HLA data

* `clone_IMGTHLA()`: Clone the _ANHIG/IMGTHLA_ repo on [GitHub](https://github.com/ANHIG/IMGTHLA) to a location specified by the option `hlatools.local_repos`. Defaults to _~/local/db/_.

* `pull_IMGTHLA()`: Send a pull request to _ANHIG/IMGTHLA_.

* `check_IMGTHLA()`: Check if your local repository is up-to-date.

### Read HLA data

A HLA gene can be read into a `HLAGene` object:

```{r}
x <- hlatools::HLAGene("DPB1")
x
```

A number of accessor functions can be used to work with these objects:

```{r}
x <- hlatools::HLAGene("DPA1")
hlatools::db_version(x)
hlatools::locusname(x)
x1 <- x[hlatools::is_complete(x)][1:10]
hlatools::allele_name(x1)
hlatools::cwd_status(x1)
hlatools::ethnicity(x1)
hlatools::g_group(x1)
hlatools::p_group(x1)
hlatools::sequences(x1)
hlatools::exon(x1, exon = 2)
```

We can access previous releases of the IPD-IMGT/HLA database:

```{r}
x <- hlatools::HLAGene("DPB1", db_version = "3.25.0")
x
hlatools::db_version(x)
hlatools::locusname(x)
x[hlatools::cwd_status(x) == "Common"][1:10]
```

These objects come with an additional set of `R6`-methods that implement an API used mostly in conjunction with the `DR2S` package:

* `x$get_closest_complete_neighbor(allele, partially = TRUE)`: Find the closest full-length allele based on the genetic distance between exon sequences.

* `x$get_reference_sequence(allele)`: Return a `BStringSet` object of a full-length reference sequence for `allele`. If `allele` is not fully known the missing stretches are taken from the allele returned by a call to `get_closest_complete_neighbor()`.

### Session Info

```{r, echo=FALSE}
devtools::session_info()
```


